kube-core/src/schema/transform_any_of.rs

```diff
@@ -0,0 +1,238 @@
+///
+/// This will return early without modifications unless:
+/// - There are exactly 2 `anyOf` subschemas.
+/// - One subschema represents the nullability (ie: it has an `enum` with a single
+///   `null` entry, and `nullable` set to true).
+///
+/// NOTE: This should work regardless of whether other hoisting has been performed or not.
+#[derive(Debug, Clone)]
+pub struct AnyOfSchemaRewriter;
+
+impl Transform for AnyOfSchemaRewriter {
+    fn transform(&mut self, transform_schema: &mut schemars::Schema) {
+        // Apply this (self) transform to all subschemas
+        schemars::transform::transform_subschemas(self, transform_schema);
+
+        let mut schema: SchemaObject = match serde_json::from_value(transform_schema.clone().to_value()).ok()
+        {
+            // TODO (@NickLarsenNZ): At this point, we are seeing duplicate `title` when printing schema as json.
+            // This is because `title` is specified under both `extensions` and `other`.
+            Some(schema) => schema,
+            None => return,
+        };
+
+        hoist_any_of_subschema_with_a_nullable_variant(&mut schema);
+
+        // Convert back to schemars::Schema
+        if let Ok(schema) = serde_json::to_value(schema) {
+            if let Ok(transformed) = serde_json::from_value(schema) {
+                *transform_schema = transformed;
+            }
+        }
+    }
+}
+
+fn hoist_any_of_subschema_with_a_nullable_variant(kube_schema: &mut SchemaObject) {
+    // Run some initial checks in case there is nothing to do
+    let SchemaObject {
+        subschemas: Some(subschemas),
+        ..
+    } = kube_schema
+    else {
+        return;
+    };
+
+    let SubschemaValidation {
+        any_of: Some(any_of),
+        one_of: None,
+    } = subschemas.deref_mut()
+    else {
+        return;
+    };
+
+    if any_of.len() != 2 {
+        return;
+    }
+
+    let entry_is_null: [bool; 2] = any_of
+        .iter()
+        .map(|schema| serde_json::to_value(schema).expect("schema should be able to convert to JSON"))
+        .map(|value| value == *NULL_SCHEMA)
+        .collect::<Vec<_>>()
+        .try_into()
+        .expect("there should be exactly 2 elements. We checked earlier");
+
+    // Get the `any_of` subschema that isn't the null entry
+    let subschema_to_hoist = match entry_is_null {
+        [true, false] => &any_of[1],
+        [false, true] => &any_of[0],
+        _ => return,
+    };
+
+    // At this point, we can be reasonably sure we need to hoist the non-null
+    // anyOf subschema up to the schema level, and unset the anyOf field.
+    // From here, anything that looks wrong will panic instead of return.
+    // TODO (@NickLarsenNZ): Return errors instead of panicking, leave panicking up to the
+    // infallible schemars::Transform
+    let Schema::Object(to_hoist) = subschema_to_hoist else {
+        panic!("the non-null anyOf subschema is a bool. That is not expected here");
```

Comment:
We should not panic in this code. This produces no linter hints, and gives no trace on which field actually failed. The only correct approach with the chosen path by `schemars` is to split code into more domain specific parts, and log errors instead of panicking here.